Add proposals/lambdas-* documents for discussion.

jemc opened the pull request:
@jasoncarr0 requested that I convert my proposal notes in #130 into a pull request of documents for easier review and threading of review comments.

Jason and anyone else are welcome to review this here and any changes I make in this PR will get later reflected in ticket #130 when I edit those comments to add that information.

jasoncarr0 commented on proposals/lambdas-06-decision-tree.md:
```diff
@@ -0,0 +1,21 @@
+
+4. "All of the arguments I need to pass into this lambda are sendable and the result I get out from it can also be sendable"
+  - if true: use `Fun'iso`
+  - if false: use `FunVal` (a.k.a `FunVal'val`)
```
Do we have any use case where taking a `box` fun could be useful for callers?

jasoncarr0 replied:
Ah no that would be guaranteeing that the best that we can give is box, so we'd have to have both refs and vals which is an enormous mess I could never see happening with lambdas

jasoncarr0 replied:
I guess if it gets stored such that it never leaves the thread but may need to be called from a box method...

jemc replied:
I suppose this is a potential use case, though I'm not sure how common it will be. I was hoping to avoid a fourth trait to avoid adding more mental load on which trait to use, but it seems this one may not be totally avoidable.

I didn't really want to consider it because it is strictly weaker at the block site and I didn't see it as having any benefits. But you're right - it is beneficial at the lambda object use site for the specific case where the lambda object user needs to store the lambda in a field and then retrieve and call it from that field in the middle of a `:fun box`, where it cannot retrieve and call a `:fun ref` on that field.

If I add nothing beyond what is already in the proposal, in such a situation the lambda user would need to specify `FunVal'val`, meaning the lambda supplier would be unnecessarily restricted in not being able to capture non-shareable values from the outer scope. I say it's unnecessary because we've established in this example that the lambda user doesn't actually need it to be sendable, so `val` is unnecessarily strong where `box` could be used.

---

So we need a lambda "Kind" that has the following properties:

Call count:
  - can be called zero or more times

Inter-scope interaction:
  - can capture variables from the outer scope
  - CANNOT rebind captured variables in the inner scope
  - CANNOT consume `iso` variables in the inner scope
  - CANNOT affect variable bindings in the outer scope
  - CANNOT raise an error into the outer scope
  - CANNOT return early, exiting the outer scope

Ref cap restrictions:
  - can capture non-shareable values from the outer scope
  - HOWEVER, all captured values are seen through a `box` lens
  - can receive non-sendable yielded values
  - can return a non-sendable yield result value

Mobility:
  - CANNOT cross an actor boundary
  - can be objectified and stored away for later

Note that the only difference from the `FunVal'val` kind is that it is able to capture non-shareable values from the outer scope. However, all the captured values would be seen through a `box` lens, which is unfortunate but necessary.

---

Perhaps we can do this without adding a fourth trait?

Maybe we could amend the third trait, which is currently defined as:

```savi
:trait val FunVal(R, P1, P2)
  :fun val call(P1, P2) R
```

and we amend it such that it looks like this:

```savi
:trait val FunRead(R, P1, P2)
  :fun box call(P1, P2) R
```

So be default `FunRead` would be `FunRead'val`, but in this less common kind of case we are discussing `FunRead'box` could potentially be used, which would behave like what we have discussed above.

I think that works, without us needing to add the cognitive load of a fourth trait.

What do you think?

jasoncarr0 replied:
I agree, I think it's incredibly rare that FunVal is needed, but it could be possible, if the function is sending not just its arguments, but itself as a whole. This just can't happen with lambdas, and for non-lambdas a somewhat cumbersome version may be able to store a val reference to itself

jasoncarr0 replied:
By agree I mean that `FunRead` defaulting to val and being the interface would be a good move.

Then we have:

Once -> FunOnce
Mutation available -> FunMut'ref in one actor, or FunMut'iso for sending
Mutation unavailable -> FunRead'box in one actor, or FunRead'val for sharing

jemc replied:
Yes, I think we've arrived at a sane hierarchy for the purpose of supporting lambdas.

Thanks for your help!

I will revise this PR and the associated comments on the other ticket

jemc commented on proposals/lambdas-04-kinds.md:
```diff
@@ -0,0 +1,203 @@
+Internal Implementation / Optimization:
+  - when objectified, it is a heap-allocated object of an anonymous one-off type, behaving like the "as if" model, and using a virtual table for dispatch when called.
+  - in some cases, virtual dispatch on the object may be possible to avoid by using automatic specialization optimizations in the late compiler stages.
+  - a second block that is tail-chained to it without accruing additional captures may be optimized to be another virtual method attached to the same runtime object
```
When talking about "tail-chained" blocks here I am referring to the aforementioned concept that was explained in the bank account transaction example, wherein a block section would end with another `...->` symbol followed by a new section in the same block. Basically, syntax that looks like this:

```savi
@foo ...> (
| x |
  # do stuff with x
  @bar ...->
| y |
  # do stuff with y
)
```

Which is syntax sugar for:
  
```savi
@foo ...> (x |
  # do stuff with x
  @bar ...-> (y |
    # do stuff with y
  )
)
```

But when we recognize either form as a tail-chain with the second async lambda's execution following the first async lambda's execution, then as long as certain conditions are met we can quietly combine them into the same heap-allocated object to avoid an extra allocation.

> Won't this need to be a new virtual table anyway?

This optimization will only be possible when the potentially-combined lambdas have different virtual table index numbers, so that they will not collide if occupied in the same table. This should always be possible in cases where the lambda signatures (their params and return type) are different/incompatible, but may not be possible in cases where the signatures are the same/compatible.

jemc commented on proposals/lambdas-01-intro.md:
```diff
@@ -0,0 +1,26 @@
+Yield blocks use a syntax of `->` to attach the yield block to a call site. Async yield blocks will use a similar syntax of `...->`, creating what is effectively a lambda as known in other languages.
```
I don't think changing a symbol associated with behaviors would be enough because an objectified lambda that is not passed to a behavior should also have the syntax that indicates "called later" to the reader.

I am open to exploring other arrow options. For the current choice what I like about it is that it is very visually distinct and evocative for the reader, even if admittedly it is a bit more laborious to type for the writer.

I don't think I'm a fan where very small visual differences in the arrow type (`->` vs `=>`) are significant and commonplace because such small elements are very easy to visually blend together with one's eyes when reading code.

jasoncarr0 replied:
that's fair, I'm not sure it's really more evocative than `->` is for lambdas, which is the norm in most other programming languages for general lambdas

jemc commented on proposals/lambdas-04-kinds.md:
```diff
@@ -0,0 +1,203 @@
+  - when objectified, it is a heap-allocated object of an anonymous one-off type, behaving like the "as if" model, and using a virtual table for dispatch when called.
+  - in some cases, virtual dispatch on the object may be possible to avoid by using automatic specialization optimizations in the late compiler stages.
+  - a second block that is tail-chained to it without accruing additional captures may be optimized to be another virtual method attached to the same runtime object
+  - a second block that is tail-chained to it WITH additional captures accrued from the first block may get the same optimization if the first block is proved to be single-use (either by being an objectified single-use or a non-objectified definition with lexical proof of not being multi-use) - the additional captures become fields which are set during the execution of the first block and proven  ...
```
I was planning to only do this optimization in cases where the virtual table indices for the two methods were different (which is possible when they have different/incompatible type signatures)

I didn't think about the possibility of reassigning the virtual table / descriptor. That seems like a good idea for cases where the virtual table indices do collide. It's definitely much cheaper than an allocation.

jasoncarr0 replied:
Oh true, yes that would work

jemc commented on proposals/lambdas-04-kinds.md:
```diff
@@ -0,0 +1,203 @@
+
+Internal Implementation / Optimization:
+  - when objectified, it is a heap-allocated object of an anonymous one-off type, behaving like the "as if" model, and using a virtual table for dispatch when called.
+  - in some cases, virtual dispatch on the object may be possible to avoid by using automatic specialization optimizations in the late compiler stages.
```
Also I think a behavior call can never be inlined by LLVM, but it could definitely be specialized by my compiler ahead of time, so that's a common example of a need for me to take action on this, assuming that LLVM can only assist with eliminating virtual dispatch on inlined content (though again, I would be quite happy to hear about me being wrong about that)

jasoncarr0 replied:
I'm going to guess that LLVM can prove that for more cases than what you've mentioned. Or to put another way, I imagine that getting any cases that LLVM doesn't cover is probably going to be more effort than is worth spending any time in the foreseeable future

jasoncarr0 replied:
What specialization would occur that couldn't be performed at the LLVM level?

jemc replied:
I think the behavior call example I mentioned above is a common use case and reliable to reproduce that I don't think LLVM will be able to devirtualize.

If I have a behavior that accepts an abstract type as a parameter, and I pass it a concrete type as the corresponding argument, any calls made with the parameter as a receiver will be virtual calls and LLVM has no way to devirtualize them, because it isn't able to inline behavior dispatch into the behavior caller, and it has no other way to track the concrete descriptor from the behavior call site into the behavior body.

By contrast in the Savi compiler I think it would not be too tricky to generate a specialized version of the behavior that accepts the concrete type as a parameter and give it a totally distinct message id. LLVM doesn't know that it's okay to do this (make up new message ids and add corresponding new clauses to the dispatch function) so it won't, but I know the semantics of the Pony runtime and how I am using it, so I can encode the knowledge of this optimization into the Savi compiler.

There are perhaps other examples where pre-LLVM specialization can do a better job than LLVM can, but that's the clearest example I can come up with at the moment.

jemc commented on proposals/lambdas-04-kinds.md:
```diff
@@ -0,0 +1,203 @@
+  - It's as if the block becomes an object (with ref cap: `val`)
+  - the object has a single function corresponding to the block body
+  - the captured variables are arguments to a `:new val` constructor
+  - within the block, the captured variables are fields, which may not be rebound because the block function is a `:fun val`.
```
I think this is covered in this new discussion comment: https://github.com/savi-lang/savi/pull/131#discussion_r691327814

jemc commented on proposals/lambdas-04-kinds.md:
```diff
@@ -0,0 +1,203 @@
+  - in some cases, virtual dispatch on the object may be possible to avoid by using automatic specialization optimizations in the late compiler stages.
+  - a second block that is tail-chained to it without accruing additional captures may be optimized to be another virtual method attached to the same runtime object
+  - a second block that is tail-chained to it WITH additional captures accrued from the first block may get the same optimization if the first block is proved to be single-use (i.e. a non-objectified definition with lexical proof of not being multi-use) - the additional captures become fields which are set during the execution of the first block and proven to be ready when they are used in the se ...
+
```
I think this is covered in this new discussion comment: https://github.com/savi-lang/savi/pull/131#discussion_r691327814

jemc commented on proposals/lambdas-05-traits.md:
```diff
@@ -0,0 +1,72 @@
+
+Notice that these traits (`FunVal`, `Fun`, `FunOnce`, respectively) mostly mirror the hierarchy of function traits in Rust (`Fn <: FnMut <: FnOnce`), except that in our type system `FunVal` cannot generally be a subtype of `Fun` because the former is `:fun val` and the latter is `:fun ref`. However `:fun box` can satisfy a `:fun ref` trait, so meeting the `Fun` trait does not strictly require a m ...
+
+We mostly make up for deficiency of `FunVal` not being a subtype of `Fun` by relying on type anonymity to allow a `FunVal`-compatible `...->` block literal to be objectified as one of the other trait types as long as it happens immediately at the creation site. That is, we can conceptually treat the immutable block as being a sub-object inside a mutable one, as long as we know this at the definit ...
```
So far I haven't defined any way of creating a raw lambda without it being passed to some function. I actually view this limitation to be a positive feature, because it forces specific typing of all lambda objects and we shouldn't have to deal with ambiguity on how a particular block should be treated.

That is, the function you pass it to will define the type of lambda it expects to receive, and then the `...->` block you pass to it will have to meet those requirements, and will be objectified with the relevant trait and ref cap.

Note that there is also not a way in the syntax to specify the parameter types of a `...->` block at the call site. This as well is expected to be inferred from looking at the signature of the function that receives and objectifies it.

I think this is a good tradeoff because it completely removes the ambiguity in typing while keeping the type information away from being a burden on the call site.

In cases where you don't actually have a need to pass it to a function immediately, you can have named stub functions that only exist to receive a `...->` block with a specific type and immediately return it. These stub functions would be named in such a way as to help the reader understand how this lambda was to be used later, such as in the following example, where `FileWatcher.change_callback` is such a stub function (a `:fun non`, callable like a static function) that only exists for the purposes of assigning a type to the block:

```savi
on_change = FileWatcher.change_callback ...-> (change_info |
  // (do something with `change_info`)
)

// ... some time later, or in another function

my_file_watcher.on_change = on_change
```

I actually think this is quite nice way of avoiding ambiguity while preserving and even enhancing readability.

jemc commented:
Updated the documents here based on the review, and updated the corresponding ticket comments in #130, so I'm going to close this PR and we can finish the discussion there.
